pkg/kubelet/images/image_manager.go

```diff
@@ -164,14 +233,37 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
 		m.backOff.Next(backOffKey, m.backOff.Clock.Now())
 
 		msg, err := evalCRIPullErr(container, imagePullResult.err)
-		return "", msg, err
+		return imageRef, msg, err
 	}
 	m.podPullingTimeRecorder.RecordImageFinishedPulling(pod.UID)
 	imagePullDuration := time.Since(startTime).Truncate(time.Millisecond)
 	m.logIt(ref, v1.EventTypeNormal, events.PulledImage, logPrefix, fmt.Sprintf("Successfully pulled image %q in %v (%v including waiting). Image size: %v bytes.",
 		container.Image, imagePullResult.pullDuration.Truncate(time.Millisecond), imagePullDuration, imagePullResult.imageSize), klog.Info)
 	metrics.ImagePullDuration.WithLabelValues(metrics.GetImageSizeBucket(imagePullResult.imageSize)).Observe(imagePullDuration.Seconds())
 	m.backOff.GC()
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
+		func() {
+			m.lock.Lock()
+			if imagePullResult.pullCredentialsHash == "" {
+				// successful pull no auth hash returned, auth was not required so we should reset the hashmap for this
+				// imageref since auth is no longer required for the local image cache, allowing use of the ImageRef
+				// by other pods if it remains cached and pull policy is PullIfNotPresent
+				delete(m.ensureSecretPulledImages, imageRef)
+			} else {
+				// store/create hashMatch map entry for auth config hash key used to pull the image
+				// for this imageref (digest)
+				digest := m.ensureSecretPulledImages[imageRef]
+				if digest == nil {
+					digest = &imagePullInfo{Auths: make(map[string]*ensuredInfo)}
+					m.ensureSecretPulledImages[imageRef] = digest
+				}
+				digest.Auths[imagePullResult.pullCredentialsHash] = &ensuredInfo{true, time.Now()}
+			}
+			defer m.lock.Unlock()
+		}()
```

Comment:
why is there a different closure and why are we deferring at the end of it? I would expect the defer to go with the Lock() above, but really I think it'd be better without the closure and with the explicit lock/unlock directly inside the if scope

Reply:
My fault. Updated.